src/py/flwr/server/superlink/fleet/grpc_rere/server_interceptor.py

```diff
@@ -172,3 +160,56 @@ def _generic_method_handler(
             request_deserializer=method_handler.request_deserializer,
             response_serializer=method_handler.response_serializer,
         )
+
+    def _verify_node_id(
+        self,
+        node_id: Optional[int],
+        request: Union[
+            DeleteNodeRequest,
+            PullTaskInsRequest,
+            PushTaskResRequest,
+            GetRunRequest,
+            PingRequest,
+        ],
+    ) -> bool:
+        if node_id is None:
+            return False
+        if isinstance(request, PushTaskResRequest):
+            if len(request.task_res_list) == 0:
+                return False
+            return request.task_res_list[0].task.producer.node_id == node_id
+        if isinstance(request, GetRunRequest):
+            return node_id in self.state.get_nodes(request.run_id)
+        return request.node.node_id == node_id
+
+    def _verify_hmac(
+        self, public_key: ec.EllipticCurvePublicKey, request: Request, hmac_value: bytes
+    ) -> bool:
+        shared_secret = generate_shared_key(self.server_private_key, public_key)
+        return verify_hmac(shared_secret, request.SerializeToString(True), hmac_value)
+
+    def _create_authenticated_node(
+        self,
+        public_key_bytes: bytes,
+        request: CreateNodeRequest,
+        context: grpc.ServicerContext,
+    ) -> CreateNodeResponse:
+        context.send_initial_metadata(
+            (
+                (
+                    _PUBLIC_KEY_HEADER,
+                    self.encoded_server_public_key,
+                ),
+            )
+        )
+
+        node_id = self.state.get_node_id(public_key_bytes)
+
+        # Handle RPC here instead of passing to default method_handler
+        if node_id is not None:
+            # State has lock already this execution is serialized
+            self.state.acknowledge_ping(node_id, request.ping_interval)
+            return CreateNodeResponse(node=Node(node_id=node_id, anonymous=False))
+
+        node_id = self.state.create_node(request.ping_interval, public_key_bytes)
+        return CreateNodeResponse(node=Node(node_id=node_id, anonymous=False))
```

Comment:
@danieljanes why should we call FleetServicer.CreateNode? The public key bytes is in metadata and FleetServicer.CreateNode has no access to metadata. The Interceptor features allows us to reroute method_handler in specific cases like this.

Reply:
Got it. In that case, does it make sense to have `CreateNode` at all? It is only being used for non-authenticated `CreateNode` requests (i.e., requests without a public key). We could consider these to be special cases and handle them *almost* the same way we handle authenticated ones.

Reply:
It makes sense to have `CreateNode`. In case of non-authenticated server, we do not instantiate an `Interceptor`. We only reroute the `CreateNode` when authentication is enabled. 

I'd like to keep the rerouting of `CreateNode` in the `Interceptor`, not moving the authentication logic to the default`method_handler`. Right now the only downside of separating it is having this two lines of code duplication in case of new client connection that was not yet assigned a node_id:

```
node_id = self.state.create_node(request.ping_interval, public_key_bytes)
return CreateNodeResponse(node=Node(node_id=node_id, anonymous=False))
```

For this duplication, we have a clear separation of authentication logic and the runtime is not affected since the default `method_handler` is not called at all.